opm/simulators/aquifers/BlackoilAquiferModel_impl.hpp

```diff
@@ -139,72 +139,32 @@ template <typename TypeTag>
-    if (comm.rank() == 0)
-        has = deck.hasKeyword("AQUCT");
-    comm.broadcast(&has, 1, 0);
-
-    if (has) {
-         if (comm.size() > 1)
-             throw std::runtime_error("Aquifers currently do not work in parallel.");
-
-        // updateConnectionIntensiveQuantities();
-        const auto& eclState = this->simulator_.vanguard().eclState();
+    if (aquifer.active()) {
+        const auto& comm = this->simulator_.vanguard().gridView().comm();
+        if (comm.size() > 1)
+            throw std::runtime_error("Aquifers currently do not work in parallel.");
 
         // Get all the carter tracy aquifer properties data and put it in aquifers vector
-        const AquiferCT aquiferct = AquiferCT(eclState.getTableManager(), deck);
-        const Aquancon aquifer_connect = Aquancon(eclState.getInputGrid(), deck);
-
         const auto& ugrid = simulator_.vanguard().grid();
-        const auto& gridView = simulator_.gridView();
-        const int number_of_cells = gridView.size(0);
-
-        cartesian_to_compressed_ = cartesianToCompressed(number_of_cells, Opm::UgGridHelpers::globalCell(ugrid));
+        const int number_of_cells = simulator_.gridView().size(0);
 
-        for (const auto& aquifer : aquiferct)
-            aquifers_CarterTracy.push_back(AquiferCarterTracy<TypeTag>(aquifer_connect[aquifer.aquiferID], cartesian_to_compressed_, this->simulator_, aquifer));
-
-    }
-    if (comm.rank() == 0)
-        has = deck.hasKeyword("AQUFETP");
-    comm.broadcast(&has, 1, 0);
+        cartesian_to_compressed_ = cartesianToCompressed(number_of_cells,
+                                                         Opm::UgGridHelpers::globalCell(ugrid));
 
-    if (has) {
-         if (comm.size() > 1)
-             throw std::runtime_error("Aquifers currently do not work in parallel.");
-
-        // updateConnectionIntensiveQuantities();
-        const auto& eclState = this->simulator_.vanguard().eclState();
-
-        // Get all the carter tracy aquifer properties data and put it in aquifers vector
-        const Aquifetp aquifetp = Aquifetp(deck);
-        const Aquancon aquifer_connect = Aquancon(eclState.getInputGrid(), deck);
-
-        const auto& ugrid = simulator_.vanguard().grid();
-        const auto& gridView = simulator_.gridView();
-        const int number_of_cells = gridView.size(0);
-
-        cartesian_to_compressed_ = cartesianToCompressed(number_of_cells, Opm::UgGridHelpers::globalCell(ugrid));
-
-        for (const auto& aquifer : aquifetp)
-            aquifers_Fetkovich.push_back(AquiferFetkovich<TypeTag>(aquifer_connect[aquifer.aquiferID], cartesian_to_compressed_, this->simulator_, aquifer));
+        const auto& connections = aquifer.connections();
+        for (const auto& aq : aquifer.ct()) {
+            AquiferCarterTracy<TypeTag> act(connections[aq.aquiferID],
+                                            cartesian_to_compressed_, this->simulator_, aq);
+            aquifers_CarterTracy.emplace_back(act);
```

Comment:
Emplace_back on a fully constructed object is useless.  This should either be
```C++
acquifers_CarterTracey.emplace_back(connections[aq.aquiferID], cartesian_to_compressed_, this->simulator_, aq);
```
or
```C++
aquifer_.push_back(std::move(act));
```
instead

Reply:
indeed. this was a complete mess, ended up with something half-way between. to inplace construction.